tabgroups/init.lua

```diff
@@ -0,0 +1,870 @@
+
+    local _tmp = {}
+    local _gv = w2groups[w].groups[group_name]
+    for i, v in grouptabs(w, group_name) do
+        if not _gv.tabs[v] then
+            _gv.tabs[v] = { tab_hit = 1, atime = os.time(), ctime = os.time() }
+        end
+        table.insert(_tmp, {v = v,
+            hits = _gv.tabs[v].tab_hit,
+            atime = _gv.tabs[v].atime,
+            mtime = _gv.tabs[v].mtime,
+            ctime = _gv.tabs[v].ctime,
+            title = v.title,
+            n = i })
+    end
+    if field and order then
+        table.sort(_tmp, function(a, b) return _sort_by_field(field, order, a, b) end)
+    end
+
+    for i, v in ipairs(_tmp) do
+        local title = '*No title*'
+        if v.title then
+            title = v.v.title
+        end
+        if v.v == active then
+            table.insert(rows, {"<b>"..((i < 10 and i..' - ') or '')..group_name.."</b>",
+                "<b>"..lousy.util.escape(title).."</b>",
+                lousy.util.escape(v.v.uri),
+                _group = group_name, _tab = v.v})
+        else
+            table.insert(rows, {((i < 10 and i..' - ') or '')..lousy.util.escape(group_name),
+                lousy.util.escape(title),
+                lousy.util.escape(v.v.uri),
+                _group = group_name, _tab = v.v})
+        end
+    end
+    return rows
+end
+
+local function _build_tabgroup_menu_grouplist(w, field, order)
+    local rows = {{ "Group name", "Number of tabs", title = true }}
+
+    local _tmp = {}
+    for g, gv in pairs(w2groups[w].groups) do
+        table.insert(_tmp, {name = g,
+            hits = gv.group_hit,
+            atime = gv.atime,
+            mtime = gv.mtime,
+            ctime = gv.ctime,
+        })
+    end
+    if field and order then
+        table.sort(_tmp, function(a, b) return _sort_by_field(field, order, a, b) end)
+    end
+    for i, g in ipairs(_tmp) do
+        if g.name == w2groups[w].active then
+            table.insert(rows, { "<b>"..((i < 10 and i..' - ') or '')..lousy.util.escape(g.name).."</b>",
+                w2groups[w].groups[g.name]._notebook:count(),
+                _group = g.name})
+        else
+            table.insert(rows, { ((i < 10 and i..' - ') or '')..lousy.util.escape(g.name),
+                w2groups[w].groups[g.name]._notebook:count(),
+                _group = g.name})
+        end
+    end
+    return rows
+end
+
+-- visual mode
+local function build_tabgroup_menu(w, expand_group)
+    local sort = settings.get_setting("tabgroups.sort_groups_by")
+    local field, order = nil, nil
+    if expand_group then
+        sort = settings.get_setting("tabgroups.sort_tabs_by")
+    end
+    if sort then
+        field, order = string.match(sort, "^%s*(%w+)%s+(%w+)")
+    end
+
+    if expand_group then
+        return _build_tabgroup_menu_grouptabs(w, expand_group, field, order)
+    else
+        return _build_tabgroup_menu_grouplist(w, field, order)
+    end
+end
+
+local _operation = nil;
```

Comment:
what if you have multiple operations going on in separate windows?

Reply:
frankly -- i never opening several different web windows :) preffer one maximized window :) that's why i also preffer tabgroups. but probably you'r right. if different windows shares same lua process then it need to be fixed

Reply:
Yep, all the UI code is a single Lua process; the usual way to do this is to have a weak table: `local operation = setmetatable({}, { __mode = "k" })` and then key by `w`.